### PR TITLE
Added loading HDR equirectangular texture (close #189)

### LINKED ISSUE #189: PBR Renderer: load environment map from mercator projection
Currently the environment map can only be loaded from the cubemap texture, which is a very rarely used format.
A more common format is a single image with mercator projection.
An example of such HDR maps can be found at https://hdri-haven.com/

Requires https://github.com/DiligentGraphics/DiligentTools/issues/216

### PR DESCRIPTION
**Pull request by MikhailGorobets**

### CODE REVIEW

File: PBR/interface/PBR_Renderer.hpp [hunk 1]
```diff
@@ -677,6 +677,38 @@ class PBR_Renderer
     void CreatePSO(PsoHashMapType& PsoHashMap, const GraphicsPipelineDesc& GraphicsDesc, const PSOKey& Key);
 
 protected:
+    enum IBL_FEATURE_FLAGS : Uint32
+    {
+        IBL_FEATURE_FLAG_NONE = 0,
+        IBL_FEATURE_FLAG_OPTIMIZE_SAMPLES,
```
Comment (TheMostDiligent) [inline PBR/interface/PBR_Renderer.hpp:683]: Initializers are missing here

File: PBR/interface/PBR_Renderer.hpp [hunk 2]
```diff
@@ -677,6 +677,38 @@ class PBR_Renderer
...
+        IBL_FEATURE_FLAG_SAMPLING_SPHERE
+    };
+
+    struct IBL_PSOKey
+    {
+        IBL_FEATURE_FLAGS FeatureFlags;
+        TEXTURE_FORMAT    RTVFormat;
```
Comment (TheMostDiligent) [inline PBR/interface/PBR_Renderer.hpp:690]: Either use const or always default-initialize struct members

File: PBR/src/PBR_Renderer.cpp [hunk 3]
```diff
@@ -480,21 +480,45 @@ void PBR_Renderer::PrecomputeCubemaps(IDeviceContext* pCtx,
...
-        float Dummy;
+    };
+
+    auto GetIBL_PSO = [&](IBL_FEATURE_FLAGS FeatureFlags, TEXTURE_FORMAT RTVFormat) -> RefCntAutoPtr<IPipelineState>& {
+        auto Iter = m_IBL_PSOCache.find(IBL_PSOKey{FeatureFlags, RTVFormat});
+        if (Iter != m_IBL_PSOCache.end())
+            return Iter->second;
+
+        auto Condition = m_IBL_PSOCache.emplace(IBL_PSOKey{FeatureFlags, RTVFormat}, RefCntAutoPtr<IPipelineState>{});
+        return Condition.first->second;
```
Comment (TheMostDiligent) [inline PBR/src/PBR_Renderer.cpp:494]: There is a much shorter version that does exactly the same:
`m_IBL_PSOCache.emplace[IBL_PSOKey{FeatureFlags, RTVFormat}]`

File: Shaders/PBR/private/ComputeIrradianceMap.psh [hunk 4]
```diff
@@ -38,14 +56,16 @@ float3 IrradianceMap(float3 N)
         float OmegaS = 1.0 / (float(g_NumSamples) * pdf);
         
         // Solid angle of 1 pixel across all cube faces
-        float OmegaP = 4.0 * PI / (6.0 * g_EnvMapDim * g_EnvMapDim);
+        float OmegaP = 4.0 * PI / (6.0 * g_EnvMapWidth * g_EnvMapHeight);
```
Comment (TheMostDiligent) [inline Shaders/PBR/private/ComputeIrradianceMap.psh:59]: For sphere map, 6.0 should be removed

File: PBR/src/PBR_Renderer.cpp [hunk 5]
```diff
@@ -480,21 +480,45 @@ void PBR_Renderer::PrecomputeCubemaps(IDeviceContext* pCtx,
...
-    if (!m_pPrecomputeIrradianceCubePSO)
+    IBL_FEATURE_FLAGS FeatureFlags = IBL_FEATURE_FLAG_NONE;
+
+    if (OptimizeSamples)
+        FeatureFlags = static_cast<IBL_FEATURE_FLAGS>(FeatureFlags | IBL_FEATURE_FLAG_OPTIMIZE_SAMPLES);
+
+    if (!pEnvironmentMap->GetTexture()->GetDesc().IsCube())
+        FeatureFlags = static_cast<IBL_FEATURE_FLAGS>(FeatureFlags | IBL_FEATURE_FLAG_SAMPLING_SPHERE);
+
+    auto pPrecomputeIrradianceCubePSO = GetIBL_PSO(FeatureFlags, m_pIrradianceCubeSRV->GetDesc().Format);
```
Comment (TheMostDiligent) [inline PBR/src/PBR_Renderer.cpp:510]: Here is a bug: the type of pPrecomputeIrradianceCubePSO will not be reference, and the PSO will never be written to cache

File: Shaders/PBR/private/ComputeIrradianceMap.psh (hunk 4 above)
Comment (MikhailGorobets) [reply Shaders/PBR/private/ComputeIrradianceMap.psh:59]: Are you sure we only need to remove this constant? I haven't looked into how the formulas for calculating mip levels are derived

Comment (TheMostDiligent) [reply Shaders/PBR/private/ComputeIrradianceMap.psh:59]: `6.0 * g_EnvMapDim * g_EnvMapDim` is the total number of pixels in the cubemap. Clearly, for the sphere map the number of pixels is `g_EnvMapDim * g_EnvMapDim`

**Pull request merged**